keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,866 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+The Probe struct contains int32 fields that specify seconds for timeouts.
+Some users would like to have timeouts less than one second.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+#### Istio
+
+TODO(???): explain this scenario, presumably around sidecar readiness?
+
+#### Knative
+
+Knative will create Pods (via Deployment) and wait for them to become `Ready`
+when handling an HTTP request from an end-user. In this case, Pod readiness
+latency has a direct impact on HTTP request latency. (In the steady state,
+Knative will re-use an existing Pod, but this situation can happen on scale-up,
+and is guaranteed to happen on scale-from-zero.)
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+An ability to specify timeouts that are less than one second.
+Add additional tests cases to the timeout test cases.
+
+? link to existing test cases ?
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+V2 API for existing objects.
+Converting fields from int32 to resource.Quantity.
+Subsecond resolution less than one millisecond.
+
+## Proposal - Plan-A
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new int32 field to existing Probe struct for timeoutSeconds that exists of timeoutMilliseconds.
+
+int32 data type is used for consistency with existing fields.
+
+Using the fully available range of an int32 with a maximum value of 2147483647 represents a value of ~24.8 days.
+
+If the Milliseconds variant of a field is set,
+*use it in preference* to the existing Seconds field,
+and **completely ignore** the value of the existing field.
+This behavior makes it opt-in on setting a non-zero Milliseconds field.
```

Comment:
If there is an existing admission control mechanism that enforces constraints on `timeoutSeconds`, does this (draft)  change break any API contract that the admission control mechanism is entitled to assume?

Reply:
@ehashman recommended in the sig node call that we forgo the alternate fields idea and move to adding a new flag to  indicate seconds should be processed as milliseconds.. etc.

great point on we will need to check for admission control impacts..